pkg/worker/worker.go

```diff
@@ -0,0 +1,244 @@
+	job, err := worker.DataStorage.Retrieve(uuid)
+	worker.DataStorage.Unlock()
+	if err != nil {
+		return errors.Wrapf(err, "error retrieving job with UUID: %v", uuid)
+	}
+	if job.UUID == googleUUID.Nil {
+		return ErrJobNotFound
+	}
+	if !job.StoppedAt.IsZero() {
+		return ErrJobAlreadyStopped
+	}
+
+	err = syscall.Kill(job.PID, syscall.SIGKILL)
+	if err != nil {
+		return errors.Wrapf(err, "error killing the process with pid: %d", job.PID)
+	}
+
+	job.StoppedAt = time.Now()
+	err = job.store()
+
+	return err
+}
+
+// Query returns information about the job.
+func (worker *Worker) Query(uuid googleUUID.UUID) (Job, error) {
+	worker.DataStorage.Lock()
+	job, err := worker.DataStorage.Retrieve(uuid)
+	worker.DataStorage.Unlock()
+	if err != nil {
+		return Job{}, errors.Wrapf(err, "error retrieving job with UUID: %v", uuid)
+	}
+	if job.UUID == googleUUID.Nil {
+		return Job{}, ErrJobNotFound
+	}
+
+	return job, nil
+}
+
+// JobStatus defines a type for job statuses.
+type JobStatus string
+
+// The constants define all possible statuses of the job.
+var (
+	StatusCreated  = JobStatus("created")
+	StatusRunning  = JobStatus("running")
+	StatusError    = JobStatus("error")
+	StatusFinished = JobStatus("finished")
+	StatusStopped  = JobStatus("stopped")
+)
+
+// Job is an abstraction around Linux process that contains logs and metadata.
+type Job struct {
+	worker     *Worker
+	Command    string
+	Status     JobStatus
+	PID        int
+	UUID       googleUUID.UUID
+	CreatedAt  time.Time
+	StartedAt  time.Time
+	StoppedAt  time.Time
+	FinishedAt time.Time
+	User       string
+	Logs       []string
+}
+
+// newJob assigns the default, required values returning new job instance.
+func newJob(worker *Worker, command string, user string) Job {
+	return Job{
+		worker:    worker,
+		Status:    StatusCreated,
+		Command:   command,
+		UUID:      googleUUID.New(),
+		CreatedAt: time.Now(),
+		User:      user,
+		Logs:      []string{},
+	}
+}
+
+// start runs the the command in a new process, listens for StdIn and StdErr
+// and logs them to the DataStorage.
+// If the DataStorage is not accessible, the errors are raised up to the Worker.Start
+// method and logged to the standard output.
+// TODO: Add stopping subprocesses of the started processes.
+// TODO: Add separation of the StdIn and StdErr.
+func (job *Job) start() error {
+	sh := os.Getenv("SHELL")
+	cmd := exec.Command(sh, "-c", job.Command)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		err = errors.Wrap(err, "error creating StdoutPipe for Cmd")
+		return job.logAndReturnError(err)
+	}
+	cmd.Stderr = cmd.Stdout
+
+	err = cmd.Start()
+	if err != nil {
+		err = errors.Wrap(err, "error starting process")
+		return job.logAndReturnError(err)
+	}
+
+	job.Status = StatusRunning
+	job.PID = cmd.Process.Pid
+	job.StartedAt = time.Now()
+	err = job.store()
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(stdout)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
```

Comment:
From `bufio.Scanner` docs: `Scanning stops unrecoverably at EOF, the first I/O error, or a token too large to fit in the buffer.`
This means that if the job outputs a line longer than [64KB](https://golang.org/pkg/bufio/#pkg-constants), the job will fail.

Use something else that supports arbitrary output and doesn't wait for a newline before storing.